File: programmers/python/level3/kakao/2021/광고삽입.py
## https://school.programmers.co.kr/learn/courses/30/lessons/72414?language=python3
def to_seconds(time):
    h,m,s = map(int, time.split(":"))
    return h*3600 + m*60+s

def to_string(time):
    time = str(time)
    if len(time) == 1:
        return "0"+time
    return time


def solution(play_time, adv_time, logs):
    play_time = to_seconds(play_time)
    adv_time = to_seconds(adv_time)
    
    times = [0] * (play_time+1)
    first_start = (play_time+1)
    last_end = 0
    for log in logs:
        start, end = log.split("-")
        start = to_seconds(start)
        end = to_seconds(end)
        times[start] += 1
        times[end] -= 1
        
        if first_start > start:
            first_start = start
        if last_end < end:
            last_end = end
        
    
    for t in range(first_start, last_end+1):
        if(t==0): continue
        times[t] = times[t] + times[t-1]
    
    ## 투 포인터 이용한 풀이 
    ans = sol_with_two_pinter(times, first_start, last_end, play_time, adv_time)
    ## 누적합 이용한 풀이 
    # ans = sol_with_prefix_sum(times, first_start, last_end, play_time, adv_time)
    
    h, left = divmod(ans, 3600)
    m, s = divmod(left, 60)
    
    return to_string(h) + ":" + to_string(m) + ":" + to_string(s)


### 투 포인터 이용한 풀이 
def sol_with_two_pinter(times, first_start, last_end, play_time, adv_time):
    start = max(0, first_start-adv_time)
    end = min(last_end+adv_time, play_time)
    
    
    s = sum(times[start:start+adv_time])
    left = start
    right = start+adv_time-1
    
    ans_cnt = s
    ans = start
    while right < end:
        s -= times[left]
        left += 1
        right += 1
        s += times[right]
        if s > ans_cnt:
            s = ans_cnt
            ans = left 

    return ans


### 누적합 이용한 풀이 
def sol_with_prefix_sum(times, first_start, last_end, play_time, adv_time):
    for t in range(first_start,  play_time +1):
        times[t] = times[t] + times[t-1]
        
    start = max(adv_time-1, first_start-adv_time)
    end = min(last_end+adv_time, play_time)
    
    ans_cnt = 0
    ans = None
    
    for end_time in range(start, end+1):
        if end_time >= adv_time:
            time = times[end_time] - times[end_time-adv_time]
        else:
            time = times[end_time]
        
        if time > ans_cnt:
            ans_cnt = time 
            ans = end_time-adv_time+1
    return ans

File: programmers/java/src/level3/kakao/kakao2021/광고삽입.java
package level3.kakao.kakao2021;

import java.util.Arrays;

//https://school.programmers.co.kr/learn/courses/30/lessons/72414?language=java
public class 광고삽입 {
    private int toSeconds(String time) {
        String[] str_times = time.split(":");
        return Integer.parseInt(str_times[0]) * 3600
                + Integer.parseInt(str_times[1]) * 60
                + Integer.parseInt(str_times[2]);
    }

    private String toStringTime(int time) {
        String strTime = String.valueOf(time);
        if(strTime.length() == 1) {
            strTime = "0" + strTime;
        }
        return strTime;
    }

    public String solution(String play_time, String adv_time, String[] logs) {
        int playTime = toSeconds(play_time);
        int advTime = toSeconds(adv_time);
        long[] times = new long[playTime + 1];

        int firstStart = playTime+1;
        int lastEnd = 0;

        for(String log: logs) {
            String[] split = log.split("-");
            int start = toSeconds(split[0]);
            int end = toSeconds(split[1]);

            times[start] += 1;
            times[end] -= 1;

            if(firstStart > start) {
                firstStart = start;
            }
            if(lastEnd < end) {
                lastEnd = end;
            }
        }


        for(int i=firstStart; i<=lastEnd; i++) {
            if(i==0) {
                continue;
            }
            times[i] = times[i] + times[i-1];
        }


        int start = Math.max(0, firstStart-advTime);
        int end = Math.min(playTime, lastEnd+advTime);

        long sum = 0;
        for(int i=start; i<start+advTime; i++) {
            sum += times[i];
        }
        int left = start;
        int right = start+advTime-1;


        long ansCnt = sum;
        int ans = start;
        while(right < end) {
            sum -= times[left];
            left += 1;
            right += 1;
            sum += times[right];
            if(sum > ansCnt) {
                sum = ansCnt;
                ans = left;
            }
        }


        int intHour = ans / 3600;
        ans %= 3600;
        int intMin = ans / 60;
        ans %= 60;

        String h = toStringTime(intHour);
        String m = toStringTime(intMin);
        String s = toStringTime(ans);

        return h+":"+m+":"+s;

    }
}
